### PR TITLE
bpf: Suppress Key Mismatch Log for Certain Maps

### LINKED ISSUE #25224: Key-size mismatch warning on 1.14 upgrades/downgrades
https://github.com/cilium/cilium/pull/23885 introduced a warning on upgrades/downgrades to/from v1.14:
```
"Key-size mismatch for BPF map" file-path=/sys/fs/bpf/tc/globals/cilium_policy_00835 new=8 old=12 subsys=bpf
```
This warning can't be avoided by users today and it is expected/non-actionable.

We need to fix that before we release v1.14.
cc @jrajahalme

### PR DESCRIPTION
**Pull request by nathanjsweet**
Some bpf maps do not affect any data loss or
require any action by operators when there is a
key size mismatch between bpf map versions.
When that is the case the warning log message
should be suppressed so as not to alarm
operators.

Fixes: #25224

### CODE REVIEW

File: pkg/bpf/bpf_linux.go [hunk 1]
```diff
@@ -52,11 +56,26 @@ func objCheck(m *ebpf.Map, path string, mapType ebpf.MapType, keySize, valueSize
 		mismatch = true
 	}
 
+	if mi, err := m.Info(); err != nil {
+		scopedLog.Error("Could not get mapinfo for map %+v", *m)
```
Comment (dylandreimerink) [inline pkg/bpf/bpf_linux.go]: The linter points it out already, but just to be sure, this should be an `Errorf` not an `Error` given the format option

File: pkg/bpf/bpf_linux.go [hunk 2]
```diff
@@ -20,6 +21,8 @@ import (
 	"github.com/cilium/cilium/pkg/spanstat"
 )
 
+var keySizeExceptions = []string{"cilium_policy"}
```
Comment (dylandreimerink) [inline pkg/bpf/bpf_linux.go]: Could you add a comment to explain when maps are exempt from the key size check? so it is clear why this code exists.

**Review by dylandreimerink — CHANGES_REQUESTED** [review]

**Review by dylandreimerink — APPROVED** [review]
:rocket:

**Comment by joestringer** [comment]
I know that @ti-mo and @jrajahalme had discussed approaches to solve this in the past, so I'll CC them for their input on the long term maintainability approach here.

**Comment by ti-mo** [comment]
This is a band-aid, and objCheck is going to be removed soon. Maps that tolerate data loss (typically maps that are write-only from the agent side like ipcache and policy) should use `Map.Recreate()` instead of `Map.OpenOrCreate()` to explicitly indicate this fact. Maybe this PR could take that approach instead?

I have patches floating around that do this for most maps, along with changing `OpenOrCreate()` to no longer implicitly recreate maps, but haven't had the time to polish them and put them up for review, ETA would be early August. This is part of the ongoing bpf.Map work.

**Comment by nathanjsweet** [comment]
A couple of things:
1. https://github.com/cilium/cilium/issues/26793 is filed to track managing this problem more effectively/comprehensively.
2. For some reason, CI ran Jenkins jobs for this PR, even though they no longer exist. The Jenkins jobs are the only failures for this CI. Marking ready-to-merge.

**Comment by nathanjsweet** [comment]
Closing. See [this issue comment](https://github.com/cilium/cilium/issues/25224#issuecomment-1633136799) for why.

**Pull request closed without merging**